ObjectsController#show: handle all Cocina::Mapper.build exceptions as 422

jmartin-sul opened the pull request:
## Why was this change made?

Handle unexpected errors from `Cocina::Mapper.build` so that an informative 422 is returned (with a ~~backtrace filtered to dor-services-app code~~full backtrace because my first attempt at filtering was hacky and this seemed useful as-is).  This will hopefully help future `Cocina::Mapper` debugging.

closes #1284

* supported by a bit more hierarchy for `Cocina::Mapper` exceptions
* `ObjectsController#json_api_error`: param for optional meta field in JSON-API error
  * `#show`: use meta field for returning backtrace
* test the new functionality
* update openapi spec for new json api error field

## How was this change tested?

unit tests

## Which documentation and/or configurations were updated?

hopefully code comments and new test suffice

jmartin-sul commented:
CI sussed out some brittle code, pushing a fix momentarily...

jcoyne commented on app/controllers/objects_controller.rb:
```diff
@@ -135,7 +140,9 @@ def json_api_error(status:, message:, title: nil)
                  'detail': message
-               }
+               }.tap do |h|
+                 h[:backtrace] = backtrace.join("\n") if backtrace
```
This field is not compliant with JSON-API (see https://jsonapi.org/format/1.1/#error-objects) maybe push it into the `detail`.

jcoyne replied:
It looks like `meta` might be another option (but you should then add it to the openapi.yml.

jmartin-sul commented:
ok @jcoyne, i switched to returning a more structured backtrace array wrapped in the `meta` field, cleaned up the test to take advantage of the structure, and attempted to update the openapi spec to match the change.  ready for another look, i think.

jcoyne commented on app/services/cocina/mapper.rb:
```diff
@@ -3,17 +3,24 @@
+    class InvalidDescMetadata < BuildError; end
+
+    # Raised on unexpected mapper failures, so .build errors can be rescued without rescuing all of StandardError
+    class UnexpectedBuildError < BuildError; end
```
Why not just rescue StandardError, since you wrap all StandardErrors with this class?

jmartin-sul replied:
i didn't want to catch errors from either of the lines above `Cocina::Mapper.build` since the ticket seemed to be asking for handling of "descriptive conversion" which i read (maybe incorrectly?) as being about the `.build` call.

happy to simplify and just catch `StandardError`.
